File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    int n;
    int[][] computers;
    boolean[] vis;
    
    void bfs(int start) {
        Queue<Integer> q = new LinkedList<>();
        q.add(start);
        
        while(!q.isEmpty()) {
            int curr = q.poll();
            
            for(int i=0; i<n; i++) {
                if(i == curr) continue;
                if(!vis[i] && computers[curr][i] == 1) {
                    vis[i] = true;
                    q.add(i);
                }
            }
        }
    }
    
    public int solution(int n, int[][] computers) {
        this.n = n;
        this.computers = computers;
        vis = new boolean[n];
        
        int answer = 0;
        
        for(int i=0; i<n; i++) {
            if(!vis[i]) {
                vis[i] = true;
                bfs(i);
                answer++;
            }
        }
        
        
        return answer;
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>

using namespace std;
int N; 
vector<vector<int>> coms;
vector<bool> vis;

void dfs(int idx) {
    if(idx == N) return;
    
    for(int i=0; i<N; i++) {
        if(!vis[i] && coms[idx][i] == 1) {
            vis[i] = true;
            dfs(i);
        }
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    N = n; coms = computers;
    vis.resize(N,false);
    
    for(int i=0; i<N; i++) {
        if(!vis[i]) {
            answer++;
            vis[i] = true;
            dfs(i);
        }
    }
    
    return answer;
}